src/vma/dev/ring_bond.h

```diff
@@ -101,51 +109,72 @@ class ring_bond_eth : public ring_bond
 {
 public:
 	ring_bond_eth(int if_index):
-		ring_bond(if_index) {}
+		ring_bond(if_index) {
+		net_device_val* p_ndev =
+				g_p_net_device_table_mgr->get_net_device_val(m_parent->get_if_index());
+		if (p_ndev) {
+			const slave_data_vector_t& slaves = p_ndev->get_slave_array();
+			for (size_t i = 0; i < slaves.size(); i++) {
+				slave_create(slaves[i]->if_index);
+			}
+		}
+	}
+
 protected:
-	virtual void    slave_create(int if_index);
-	virtual void    slave_destroy(int if_index);
+	virtual void slave_create(int if_index);
 };
 
-class ring_bond_eth_netvsc : public ring_bond_eth
+class ring_bond_ib : public ring_bond
 {
 public:
-	ring_bond_eth_netvsc(int if_index);
-	virtual ~ring_bond_eth_netvsc();
-
-	virtual bool attach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink* sink);
-	virtual bool detach_flow(flow_tuple& flow_spec_5t, pkt_rcvr_sink* sink);
-	inline void set_tap_data_available() { m_tap_data_available = true;};
-	void create_resources() {
-		// Initialize rx buffer poll
-		request_more_rx_buffers();
-		m_rx_pool.set_id("ring_bond_eth_netvsc (%p) : m_rx_pool", this);
+	ring_bond_ib(int if_index):
+		ring_bond(if_index) {
+		net_device_val* p_ndev =
+				g_p_net_device_table_mgr->get_net_device_val(m_parent->get_if_index());
+		if (p_ndev) {
+			const slave_data_vector_t& slaves = p_ndev->get_slave_array();
+			for (size_t i = 0; i < slaves.size(); i++) {
+				slave_create(slaves[i]->if_index);
+			}
+		}
 	}
 
-private:
-
-	int poll_and_process_element_tap_rx(void* pv_fd_ready_array = NULL);
-	bool request_more_rx_buffers();
-	inline void prepare_flow_message(vma_msg_flow& data, flow_tuple& flow_spec_5t, msg_flow_t flow_action);
-
-	ring_stats_t	m_ring_stat;
-	descq_t         m_rx_pool;
-	const uint32_t  m_sysvar_qp_compensation_level;
-	int             m_netvsc_idx;
-	int             m_tap_idx;
-	int             m_tap_fd;
-	bool            m_tap_data_available;
+protected:
+	virtual void slave_create(int if_index);
 };
 
-
-class ring_bond_ib : public ring_bond
+class ring_bond_netvsc : public ring_bond
 {
 public:
-	ring_bond_ib(int if_index):
-		ring_bond(if_index) {}
+	ring_bond_netvsc(int if_index):
+		ring_bond(if_index) {
+		net_device_val* p_ndev =
+				g_p_net_device_table_mgr->get_net_device_val(m_parent->get_if_index());
+
+		m_vf_ring = NULL;
+		m_tap_ring = NULL;
+		if (p_ndev) {
+			const slave_data_vector_t& slaves = p_ndev->get_slave_array();
+			for (size_t i = 0; i < slaves.size(); i++) {
+				slave_create(slaves[i]->if_index);
+			}
+			if (m_tap_ring && m_vf_ring) {
+				ring_tap* p_ring_tap = dynamic_cast<ring_tap*>(m_tap_ring);
+				if (p_ring_tap) {
+					p_ring_tap->set_vf_ring(m_vf_ring);
+				}
+			}
+		}
+	}
+
+	inline bool is_vf_mode() { return (bool)m_vf_ring; }
+
 protected:
-	virtual void    slave_create(int if_index);
-	virtual void    slave_destroy(int if_index);
+	virtual void slave_create(int if_index);
+
+private:
+	ring_slave*      m_vf_ring;
+	ring_slave*      m_tap_ring;
```

Comment:
why do we need to save the pointer here and not in the slaves array ? why not ring_simple* and ring_tap* ?